In [8]:
import os
import pandas as pd
import sqlalchemy
from dotenv import load_dotenv

load_dotenv()

db_user = os.getenv("PG_USER")
db_pass = os.getenv("PG_PASSWORD")
db_host = os.getenv("PG_HOST")
db_port = "5432"
db_name = os.getenv("PG_DB")

connection_str = f"postgresql+psycopg2://{db_user}:{db_pass}@{db_host}:{db_port}/{db_name}"
engine = sqlalchemy.create_engine(connection_str)

## Descriptive Analysis

**Business Question:**  
What is the average rent and square footage of Craigslist listings grouped by number of bedrooms?

In [9]:
sql_query = '''
WITH valid_rentals AS (
    SELECT *
    FROM raw.craigslist_data
    WHERE price_usd IS NOT NULL AND beds IS NOT NULL
)

SELECT 
    beds,
    COUNT(*) AS listings,
    ROUND(AVG(price_usd)::NUMERIC, 0) AS avg_rent,
    ROUND(AVG(sqft)::NUMERIC, 0) AS avg_sqft
FROM valid_rentals
GROUP BY beds
ORDER BY beds;
'''

In [10]:
pd.set_option('display.max_rows', None)
df = pd.read_sql(sql_query, engine)
df

beds  listings  avg_rent  avg_sqft
0   1.0        23    2525.0     693.0
1   2.0        14    2990.0    1037.0
2   3.0         4    3175.0    1323.0

**Insight:**  
Craigslist listings show a clear rent-size progression: as bedroom count increases, so does rent and square footage. 1-bed units rent for ~$2,990 on average, with 2-bed units reaching ~$3,175 — a 6% increase for ~28% more space.

**Recommendation:**  
From a tenant's perspective, 2-bed units offer better value per square foot. For investors or landlords, 1-bed units may yield more revenue per sq ft and require less upfront investment.

**Prediction:**  
As housing demand grows and affordability tightens, 1–2 bedroom rentals will continue to dominate listings and drive consistent cash flow in dense urban markets like Los Angeles.

## Diagnostic Analysis

**Business Question:**  
Which Craigslist listings offer the lowest rent per square foot, and how does that vary across units?

In [11]:
sql_query = '''
WITH rent_per_sqft AS (
    SELECT 
        address,
        price_usd,
        sqft,
        beds,
        ROUND((price_usd / NULLIF(sqft, 0))::NUMERIC, 2) AS rent_per_sqft
    FROM raw.craigslist_data
    WHERE price_usd IS NOT NULL AND sqft IS NOT NULL
)

SELECT 
    address,
    price_usd,
    sqft,
    beds,
    rent_per_sqft
FROM rent_per_sqft
ORDER BY rent_per_sqft
LIMIT 10;
'''

In [12]:
df = pd.read_sql(sql_query, engine)
df

address  price_usd    sqft  beds  \
0                             700 W Ave I       1800   895.0   2.0   
1                         1849 Wilcox Ave       2900  1400.0   3.0   
2            6944 Lurline Ave Canoga Park       2950  1342.0   3.0   
3                    17621 W. Pauline Ct.       2829  1154.0   2.0   
4  324 S Coronado St near West 4th Street       3950  1500.0   3.0   
5                      28923 N. Prarie Ln       2599   971.0   2.0   
6                       27303 N. Sara St.       2899  1049.0   3.0   
7                 2600 South Azusa Avenue       2584   929.0   2.0   
8                         Los Angeles, CA       3413  1170.0   2.0   
9                         1133 Justin Ave       2890   975.0   2.0   

   rent_per_sqft  
0           2.01  
1           2.07  
2           2.20  
3           2.45  
4           2.63  
5           2.68  
6           2.76  
7           2.78  
8           2.92  
9           2.96

**Insight:**  
The top 10 Craigslist listings by lowest rent per square foot all fall below $3.00/sqft, with most in the $2.00–$2.80 range. These listings typically offer 2–3 bedrooms and decent overall size, indicating high affordability for tenants.

**Recommendation:**  
These types of listings may be ideal for cost-conscious renters seeking more space without exceeding typical rent budgets. From a landlord or investor perspective, these units may have room for price optimization or be undervalued relative to market norms.

**Prediction:**  
As affordability pressures persist, listings offering <$3.00/sqft rent are likely to attract high demand, particularly in mid-tier submarkets. Investors who can price just above this threshold may balance strong occupancy with improved returns.

## Bonus Query: Fulfilling JOIN + Window Function Requirement

**Business Question:**  
Which Craigslist listings offer the best rent-per-square-foot value within each bedroom category?

This query joins the full listing set with a filtered subset of valid entries and ranks rent efficiency using a `RANK()` window function.

In [13]:
sql_query = '''
WITH valid_listings AS (
    SELECT *
    FROM raw.craigslist_data
    WHERE price_usd IS NOT NULL AND sqft IS NOT NULL AND beds IS NOT NULL
),

rent_value AS (
    SELECT 
        address,
        price_usd,
        sqft,
        beds,
        ROUND((price_usd / NULLIF(sqft, 0))::NUMERIC, 2) AS rent_per_sqft
    FROM valid_listings
),

ranked_rent AS (
    SELECT *,
           RANK() OVER (PARTITION BY beds ORDER BY rent_per_sqft ASC) AS rank_within_beds
    FROM rent_value
)

SELECT *
FROM ranked_rent
WHERE rank_within_beds <= 2
ORDER BY beds, rent_per_sqft;
'''


In [14]:
df = pd.read_sql(sql_query, engine)
df

address  price_usd    sqft  beds  \
0  222 1/2 Pomona Ave. near 2nd Street       2895   975.0   1.0   
1                    8762 Etiwanda Ave       1890   630.0   1.0   
2                          700 W Ave I       1800   895.0   2.0   
3                 17621 W. Pauline Ct.       2829  1154.0   2.0   
4                      1849 Wilcox Ave       2900  1400.0   3.0   
5         6944 Lurline Ave Canoga Park       2950  1342.0   3.0   

   rent_per_sqft  rank_within_beds  
0           2.97                 1  
1           3.00                 2  
2           2.01                 1  
3           2.45                 2  
4           2.07                 1  
5           2.20                 2

**Insight:**  
Across each bedroom segment, listings with the lowest rent per square foot range from ~$2.00–$3.00. These units provide strong value for renters and suggest opportunities for landlords to capture consistent occupancy with competitive pricing.

**Recommendation:**  
Consider prioritizing units that fall in the top two rent-efficiency ranks per bedroom count. These listings offer more space per dollar, which may support faster leasing and higher tenant satisfaction.

**Prediction:**  
As renters seek better cost-per-square-foot value, properties like these will become more competitive, especially in high-rent metro areas like Los Angeles.